File: Python_Code/TTSVD.py
import numpy as np
import numpy.linalg as la
import tensorly as tl
import tensorly.tenalg.proximal as prox

#import julia
#julia.install()
#from julia.api import Julia
#jl=Julia(compiled_modules=False)
#from julia import ITensorTCI
#ii = ITensorTCI.interpolative
#M = [[1,2],[3,4]]
#ii(M,1)


def TTSVD(tensorX: tl.tensor, r_max: int, eps: float) -> list[tl.tensor]:
    shape = tensorX.shape  # Get the shape of input tensor: [n1, n2, ..., nd]
    dim = len(shape)       # Get the number of dimension
    delta = (eps / np.sqrt(dim - 1)) * tl.norm(tensorX, 2)  # Truncation parameter
    W = tensorX  # Copy tensor X -> W
    nbar = W.size   # Total size of W
    r = 1              # Rank r
    ttList = []        # list storing tt factors
    iterlist = list(range(1, dim))  # Create iteration list: 1, 2, ..., d-1
    iterlist.reverse()              # Reverse the iteration list: d-1, ..., 1 
    for i in iterlist:
        W = tl.reshape(W, [int(nbar / r / shape[i]), int(r * shape[i])])  # Reshape W
        U, S, Vh = la.svd(W)  # SVD of W matrix
        # Compute rank r
        s = 0
        j = S.size 
        while s <= delta * delta:  # r_delta_i = min(j:sigma_j+1^2 + sigma_j+2^2 + ... <= delta^2)
            j -= 1
            s += S[j] * S[j]
        j += 1
        ri = min(j, r_max)  # r_i-1 = min(r_max, r_delta_i)
        Ti = tl.reshape(Vh[0:ri, :], [ri, shape[i], r])
        nbar = int(nbar * ri / shape[i] / r)  # New total size of W
        r = ri  # Renewal r
        W = U[:, 0:ri] @ np.diag(S[0:ri])  # W = U[..] * S[..]
        ttList.append(Ti)  # Append new factor
    T1 = tl.reshape(W, [1, shape[0], r])
    ttList.append(T1)    
    return ttList

I = 3
J = 4
K = 5
M = 2
N = 3
T = np.zeros([I,J,K,M,N])
for i in range(I):
    for j in range(J):
        for k in range(K):
            for m in range(M):
                for n in range(N):
                    T[i,j,k,m,n] = i-j+k-m+n
factors = TTSVD(T, 5, 1e-4)
factors.reverse()
reconT = tl.tt_to_tensor(factors)
print(tl.norm(T-reconT, 2)/tl.norm(T, 2))

pass

'''
rng = tl.check_random_state(10)  # Fix the random seed for reproducibility
shapeX = [5, 3, 7, 8, 4]         # Dimension of the tensor X                         
tensorX = T.tensor(rng.normal(size=shapeX, loc=0, scale=1), dtype='float32')  # Generate a random tensor X

ttFactors = TTSVD(tensorX, 10, 1e-3)
ttFactors.reverse()
reconX = tl.tt_to_tensor(ttFactors)
error = tl.norm(tensorX - reconX, 2)/tl.norm(tensorX, 2)
print(error)
'''

File: C_Code/src/algorithm/ttsvd.cpp
#include "header.h"

std::vector<tblis::tensor<double>> TT_SVD_dense(tblis::tensor<double> tensor, int r_max, double eps)
{    
    // Initial setting
    auto shape = tensor.lengths(); // Get the shape of the input tensor: [n1, n2, ..., nd]
    int dim = shape.size();        // Get the number of dimension d
    double delta = (eps / std::sqrt(dim - 1)) * util::Norm(tensor, 2);  // Truncation parameter
    auto W = tensor;               // Copy tensor to W 
    auto nbar = util::GetSize(W);  // Total size of W
    int r = 1;                     // Rank r
    std::vector<tblis::tensor<double>> ttList;  // List storing TT factors
    bool verbose = 0;
    // TT-SVD iteration. Iterate from d-1 to 1
    for (int i = dim-1; i>0; i--) {
        int row = nbar / r / shape[i];       // Reshape row
        int col = r * shape[i];              // Reshape column
        // Initialize SVD factors
        double* U = new double[row * row];   
        double* Vh = new double[col * col];
        double* S = new double[std::min(row, col)];
        // Dense double SVD (Lapacke)
        dSVD(W.data(), row, col, S, U, Vh);
        // Cutoff selection
        double s = 0;
        int j = std::min(row, col);
        while (s <= delta * delta) {
            j -= 1;
            s += S[j] * S[j];
        }
        j += 1;
        int ri = std::min(j, r_max);
        // Print the low rank approximation error
        if (verbose) {
            std::cout << "Iter" << i << std::endl;
            // TODO...
        }
        // Form a new TT factor
        tblis::tensor<double> factor({ri, shape[i], r});
        if (i == dim - 1) factor.resize({ri, shape[i]});
        std::copy(Vh, Vh + ri * shape[i] * r, factor.data());
        nbar = nbar * ri / shape[i] / r;  // New total size of W
        r = ri;
        // To be polished/modified 
        W.resize({row, ri});
        for (int i = 0; i<row; ++i)
            for (int j=0; j<ri; ++j) {
                W(i,j) = U[i * row + j] * S[j];
            }
        // Append the factor to the ttList
        ttList.push_back(factor);    
        delete[] U;
        delete[] Vh;
        delete[] S;
    }
    // Append the last factor 
    ttList.push_back(W);
    std::reverse(ttList.begin(), ttList.end());
    return ttList;
}